### PR TITLE
comment out code that removes trailing slash as per 5823.

### PR DESCRIPTION
**Pull request by src-r-r**
Signed-off-by: Jordan Hewitt <633901+src-r-r@users.noreply.github.com>

### Pull Request Checklist

* [X] Pull request is based on the develop branch
* [X] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [X] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [X] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

### CODE REVIEW

**Comment by callahad** [comment]
We actually title newsfragments after the PR, not the issue the fix, so a `git mv changelog.d/{5823,11224}.bugfix` should get the CI back underway :)

**Comment by src-r-r** [comment]
Great, thank you!

**Comment by dklimpel** [comment]
I am not sure if modify the configuration is the right direction.
IMO it should be some logic in `WellKnownBuilder` to be sure that there is no trailing slash.
https://github.com/matrix-org/synapse/blob/e320f5dba32f5b5818b6d5a0059ae388430b9a72/synapse/rest/well_known.py#L31-L47

**Comment by src-r-r** [comment]
> I am not sure if modify the configuration is the right direction. IMO it should be some logic in `WellKnownBuilder` to be sure that there is no trailing slash.
> 
> https://github.com/matrix-org/synapse/blob/e320f5dba32f5b5818b6d5a0059ae388430b9a72/synapse/rest/well_known.py#L31-L47

Gotcha. I'll modify it then. I'm not doubting you; I'm just curious--what's the reasoning?

**Comment by dklimpel** [comment]
The reason is that `public_baseurl` is used many times in code. With your change the tests fail. You have to verify every position in code where `public_baseurl` is used if this change has an impact.
Perhaps there are places where the trailing slash is necessary.

File: synapse/config/server.py [hunk 1]
```diff
@@ -264,9 +264,10 @@ def read_config(self, config, **kwargs):
         self.use_frozen_dicts = config.get("use_frozen_dicts", False)
 
         self.public_baseurl = config.get("public_baseurl")
-        if self.public_baseurl is not None:
-            if self.public_baseurl[-1] != "/":
-                self.public_baseurl += "/"
+        # Commented out due to https://github.com/matrix-org/synapse/issues/8308
+        # if self.public_baseurl is not None:
+        #     if self.public_baseurl[-1] != "/":
+        #         self.public_baseurl += "/"
```
Comment (richvdh) [inline synapse/config/server.py:270]: Could you explain how this is related to https://github.com/matrix-org/synapse/issues/8308?

**Comment by richvdh** [comment]
related to #5823

**Comment by richvdh** [comment]
> I am not sure if modify the configuration is the right direction.
> IMO it should be some logic in WellKnownBuilder to be sure that there is no trailing slash.

I'm not necessarily sure I agree with this: it seems a bit silly to explicitly add a trailing slash in `read_config`, and then strip it out again in `WellKnownBuilder`.

But for sure, if we change `read_config` in this way, we'll have to update the many places in the codebase that expect `public_baseurl` to have a trailing slash, which sounds like a lot of work.

**Comment by ShadowJonathan** [comment]
I think the commit history of this PR is a bit borked, i'd recommend looking at how to rebase your commits until the current main branch.

**Comment by src-r-r** [comment]
> I think the commit history of this PR is a bit borked, i'd recommend looking at how to rebase your commits until the current main branch.

I did do a rebase with develop. Should I do a rebase with master?

**Comment by ShadowJonathan** [comment]
No, hold on, something else weird is happening here then

**Comment by DMRobertson** [comment]
This passed most checks, but the sytest integration tests aren't happy. The failures aren't particularly clear, regrettably

From a quick look, I don't think the `needs_all` call here succeeds.

https://github.com/matrix-org/sytest/blob/3125282b7ac6de20bc8b2bcf07ca663f33250131/tests/10apidoc/13ui-auth.pl#L140-L150

I can't see what the failure is though. Maybe we need more `log_if_fail`? Perhaps @richvdh could advise?

**Review by DMRobertson — CHANGES_REQUESTED** [review]
Thanks for bearing with us on this one.

There's quite a few things in these changes that I don't follow, and in particular some config handling code that seems to have been deleted. There might be some confusion between the server and client well-known endpoints too?

Additionally, I think it might be clearer in places to use `str.rstrip` to remove trailing slashes, rather than . (I sympathise with your comment on #5823 about urljoin --- I'd be fine with that too, but painful to implement it across the whole source tree.)

Taking a step back, the plan as I as I understand it is:

- when ingesting config, remove any trailing slashes from the end of the URL.
- keep the trimmed publicbase_url around in memory
- when serving the .well-known/matrix/server file, use the trimmed version
- fix up any remaining places which implicitly were expecting the baseurl config to end in a forward slash
   -  by adding an explicit '/' to the path we build

**Comment by callahad** [comment]
I think we're going around in circles a bit because #5823 superficially _appeared_ simple, but in actually attempting to resolve it, @src-r-r has unearthed a bunch of conflicting / contradictory / undefined behavior.

I suspect it would be a worthwhile effort for someone on the Synapse team to write a new issue to replace #5823, which more completely enumerates the problem, solution, and consequences thereof.

Regardless, it seems clear to me that clients should (must?) be robust against base_url's both with _and_ without trailing slashes, doubly so since matrix.org is inconsistent with Synapse's default behavior.

**Comment by callahad** [comment]
@src-r-r Per discussion on the Matrix Spec https://github.com/matrix-org/matrix-doc/issues/3465#issuecomment-980449599 we've ultimately decided that mandating _either_ that trailing slashes are stripped _or_ that they're added is fraught with peril, and that the only practically viable path forward is to explicitly instruct clients to handle both cases. This has now become part of the Matrix spec, rendering this pull request moot.

I'm terribly sorry that the outcome of your hard work is not __✅ Merged!__, but in many ways this is a much more valuable result: your wrangling with this issue catalyzed improvements to the Matrix Spec in a way that will make the ecosystem itself much more robust. We wouldn't have gotten there without your involvement in this pull request. Thank you.

**Comment by src-r-r** [comment]
Thank you @callahad .

**Pull request closed without merging**